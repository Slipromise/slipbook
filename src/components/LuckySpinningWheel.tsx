import React, { useCallback, useMemo, useState } from "react";
import styles from "@/styles/components/LuckySpinningWheel.module.scss";

type Props = {};

// TODO: 由Props來控制(結果、數量)
function LuckySpinningWheel({}: Props) {
  const [angle, setAngle] = useState(() => 0);

  const onSpin = useCallback(() => {
    setAngle((pre) => pre + Math.ceil(Math.random() * 3600));
  }, []);

  const wheelStyles = useMemo(() => {
    return {
      outside: {
        transform: `rotate(${angle}deg)`,
      },
      inside: {
        transform: `rotate(-${angle}deg)`,
      },
    };
  }, [angle]);

  return (
    <div className={styles.container}>
      <div className={styles.wheel} style={wheelStyles.outside}>
        <span style={{ "--i": 0 } as React.CSSProperties}></span>
        <span style={{ "--i": 1 } as React.CSSProperties}></span>
        <span style={{ "--i": 2 } as React.CSSProperties}></span>
        <span style={{ "--i": 3 } as React.CSSProperties}></span>
        <div className={styles.numbers}>
          <b style={{ "--i": 0 } as React.CSSProperties}>5</b>
          <b style={{ "--i": 1 } as React.CSSProperties}>1</b>
          <b style={{ "--i": 2 } as React.CSSProperties}>4</b>
          <b style={{ "--i": 3 } as React.CSSProperties}>6</b>
          <b style={{ "--i": 4 } as React.CSSProperties}>2</b>
          <b style={{ "--i": 5 } as React.CSSProperties}>8</b>
          <b style={{ "--i": 6 } as React.CSSProperties}>7</b>
          <b style={{ "--i": 7 } as React.CSSProperties}>3</b>
        </div>
      </div>
      <div
        className={`${styles.wheel} ${styles.inner}`}
        style={wheelStyles.inside}
      >
        <span style={{ "--i": 0 } as React.CSSProperties}></span>
        <span style={{ "--i": 1 } as React.CSSProperties}></span>
        <span style={{ "--i": 2 } as React.CSSProperties}></span>
        <span style={{ "--i": 3 } as React.CSSProperties}></span>
        <div className={styles.numbers}>
          <b style={{ "--i": 0 } as React.CSSProperties}>5</b>
          <b style={{ "--i": 1 } as React.CSSProperties}>1</b>
          <b style={{ "--i": 2 } as React.CSSProperties}>4</b>
          <b style={{ "--i": 3 } as React.CSSProperties}>6</b>
          <b style={{ "--i": 4 } as React.CSSProperties}>2</b>
          <b style={{ "--i": 5 } as React.CSSProperties}>8</b>
          <b style={{ "--i": 6 } as React.CSSProperties}>7</b>
          <b style={{ "--i": 7 } as React.CSSProperties}>3</b>
        </div>
      </div>
      <div className={styles.spinBtn} onClick={onSpin}>
        Spin
      </div>
    </div>
  );
}

export default LuckySpinningWheel;
