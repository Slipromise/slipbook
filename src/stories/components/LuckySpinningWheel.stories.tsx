import LuckySpinningWheel from "@/components/LuckySpinningWheel";
import type { Meta, StoryObj } from "@storybook/react";
import React from "react";
import { useArgs } from "@storybook/preview-api";
import _ from "lodash";

const meta: Meta<typeof LuckySpinningWheel> = {
  component: LuckySpinningWheel,
  tags: ["autodocs"],
};

export default meta;

type Story = StoryObj<typeof LuckySpinningWheel>;

export const Normal: Story = {
  args: {
    items: [
      { label: "🍔", value: 1 },
      { label: "❤️", value: 2 },
      { label: "🌷", value: 3 },
      { label: "⚽", value: 4 },
      { label: "🚘", value: 5 },
      { label: "🔧", value: 6 },
      { label: "🎙️", value: 7 },
      { label: "📱", value: 8 },
      { label: "🗿", value: 9 },
      { label: "✂️", value: 10 },
    ],
    result: 1,
    onSpin: () => {},
  },
  render: function Render(
    args: React.ComponentProps<typeof LuckySpinningWheel>
  ) {
    const [{ result }, updateArgs] = useArgs();
    const { onSpin, ...props } = args;

    function onWrapSpin() {
      onSpin();
      updateArgs({ result: undefined });
      setTimeout(() => {
        updateArgs({ result: _.sample(args.items)?.value });
      }, 3000);
    }

    return (
      <LuckySpinningWheel {...props} onSpin={onWrapSpin} result={result} />
    );
  },
};

export const Reward: Story = {
  args: { ...Normal.args, result: 4 },
};

export const Spinning: Story = {
  args: { ...Normal.args, result: undefined },
};
