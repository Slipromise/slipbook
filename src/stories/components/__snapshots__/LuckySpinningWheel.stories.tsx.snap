// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stories/components/LuckySpinningWheel Normal smoke-test 1`] = `
<div class="LuckySpinningWheel_container___4z4h">
  <div class="LuckySpinningWheel_wheel__O3DCn"
       style="transform: rotate(0deg);"
  >
    <span style="--i: 0;">
    </span>
    <span style="--i: 1;">
    </span>
    <span style="--i: 2;">
    </span>
    <span style="--i: 3;">
    </span>
    <div class="LuckySpinningWheel_numbers__BUI3w">
      <b style="--i: 0;">
        5
      </b>
      <b style="--i: 1;">
        1
      </b>
      <b style="--i: 2;">
        4
      </b>
      <b style="--i: 3;">
        6
      </b>
      <b style="--i: 4;">
        2
      </b>
      <b style="--i: 5;">
        8
      </b>
      <b style="--i: 6;">
        7
      </b>
      <b style="--i: 7;">
        3
      </b>
    </div>
  </div>
  <div class="LuckySpinningWheel_wheel__O3DCn LuckySpinningWheel_inner__9nS3E"
       style="transform: rotate(0deg);"
  >
    <span style="--i: 0;">
    </span>
    <span style="--i: 1;">
    </span>
    <span style="--i: 2;">
    </span>
    <span style="--i: 3;">
    </span>
    <div class="LuckySpinningWheel_numbers__BUI3w">
      <b style="--i: 0;">
        5
      </b>
      <b style="--i: 1;">
        1
      </b>
      <b style="--i: 2;">
        4
      </b>
      <b style="--i: 3;">
        6
      </b>
      <b style="--i: 4;">
        2
      </b>
      <b style="--i: 5;">
        8
      </b>
      <b style="--i: 6;">
        7
      </b>
      <b style="--i: 7;">
        3
      </b>
    </div>
  </div>
  <div class="LuckySpinningWheel_spinBtn__P9qub">
    Spin
  </div>
</div>
`;
